### ISSUE TITLE
audio, video elements cause timeout in jest even with audio/video rules disabled

### ISSUE THREAD

**Issue by mohanraj-r** [dequelabs/axe-core#2528]
Expectation: audio, video elements are skipped processed when corresponding rules are disabled and tests do not timeout.

Actual: tests timeout even with audio/video rules disabled.

Motivation: support skipping of audio/video elements when axe is used with jest/jsdom.

<pre><code>
axe-core version: 4.0.2
jest: 26.4.2

Browser and Assistive Technology versions

For Tooling issues:
- Node version: v12.18.4
- Platform: Mac
</code></pre>

When the DOM contains a `<audio>` or `<video>` element, jest tests with axe timeout with error:
```
: Timeout - Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Timeout 
- Async callback was not invoked within the 5000 ms timeout specified by jest.setTimeout.Error:
```

Test HTML:
```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Test Page</title>
    </head>
    <body>
        <div role="main" id="host">
            <h1>This is a test</h1>
            <p>This is a test page with no violations</p>
            <audio src="https://file-examples-com.github.io/uploads/2017/11/file_example_MP3_700KB.mp3">Audio test</audio>
            <video src="https://file-examples-com.github.io/uploads/2020/03/file_example_WEBM_480_900KB.webm">Video test</video>
        </div>
    </body>
</html>
```

Axe config with preload disabled and audio/video rules disabled:
```
{
        preload: false,
        rules: {
            'color-contrast': { enabled: false },
            'audio-caption': { enabled: false },
            'video-caption': { enabled: false },
            'no-autoplay-audio': { enabled: false },
        }
```

The timeout happens when the `src` attribute is populated either with a non-empty URL (valid or invalid). When the `src` attribute is empty, timeout doesn't happen:

```html
<audio src="">Audio test</audio>
<video src="">Video test</video>
```

Increasing the jest timeout from 5000ms to 50000ms solves the timeout issue - but isn't practical for unit tests. The following warning gets printed on console with increased timeout (even with `preload` set to `false`):
```
console.warn
      Couldn't load preload assets:  Error: Preload assets timed out.
          at node_modules/axe-core/axe.js:14903:27
          at Timeout.task [as _onTimeout] (node_modules/jsdom/lib/jsdom/browser/Window.js:391:19)
          at listOnTimeout (internal/timers.js:549:17)
          at processTimers (internal/timers.js:492:7)
```

* Does axe process the content of audio/video?
* Why does the timeout still occur with the audio/video rules disabled? Shouldn't disabling those rules result in audio/video elements being skipped?

**Comment by mohanraj-r** [comment]
Can you also please clarify the purpose of `preload`.

[API doc](https://github.com/dequelabs/axe-core/blob/develop/doc/API.md#preload-configuration-details) states "Any additional assets (eg: cssom) to preload before running rules."

But was `preload` incorporated for performance reasons? Where does the `preload` step sit in terms of DOM lifecycle and axe analyze action?

**Comment by straker** [comment]
The `preload` config is suppose to signal a rule that requires loading the CCSOM, and should only be part of  some experimental rules (`css-orientation-lock` and `no-autoplay-audio`). Unless you've turned on experimental rules I wouldn't expect you to run into preload issues.

**Comment by mohanraj-r** [comment]
Thanks @straker 
What are the downsides to disabling preload ?
Would axe analyze perform slower when there are assets to be fetched ?

**Comment by straker** [comment]
Disabling preload should enable faster axe runs as it doesn't have to download all the CSS files and parse the CCSOM.

**Comment by mohanraj-r** [comment]
Thanks for the clarifications @straker 
Will close the issue for now

**Comment by tniezurawski** [comment]
@straker I've just got the same issue and ended up with disabling `no-autoplay-audio` but that rule is no longer experimental AFAIK.

Could you shed some light on "preload issues"? Why does it happen? 
Would switching `preload` off make these rules not work at all or this is some kind of optimization?

**Comment by straker** [comment]
If I recall correctly, `no-autoplay-audio` uses `preload` to ensure the audio / video element is fully loaded before trying to determine if the element is autoplaying and over 3s long. If you turned it off I believe the rule would still run, but it may not run for all audio / video elements if that element is not fully loaded.

Is the site you're running into this problem public and something I could run some tests on?

**Comment by straker** [comment]
From the code it looks like the `preload` function waits for every audio / video element to fully load before running tests. So if there's an element that isn't loading I could see it timing out. So if you'd be willing to run some tests, here's what I'd be looking for:

* What are you using to run axe? Are you using axe-core directly or using one of our APIs (Puppeteer, Playwright, etc.)?
* If running in a browser, I'd run axe directly using `await axe.run()` and look at the [Chrome Flame Chart](https://developer.chrome.com/docs/devtools/performance/reference/) to see if there is a particular function within axe-core that is slow. You could even save the recording and upload it for me to look at if you find it doesn't contain any private facing data. If you'd rather email me instead of post, I can give you my email in a private message.
* I'd also run axe with our built in performance tracking `await axe.run({ performanceTimer: true })`. These just log how long parts of the code took, so you posting them here would be helpful as well. 
* If you could tell me how many audio and video elements there are on the page, as well as if there are any iframes, that'd help narrow down if it's a problem with there being lots of nodes or just a few (`document.querySelectorAll('iframe, frame').length`, `document.querySelectorAll('audio[autoplay], video[autoplay]').length`).

**Comment by Waazer** [comment]
Robin kredittkort Robin Bitcoin åpen Kebyr Hvor er jeg hvor bra

**Comment by tniezurawski** [comment]
> [...] Do any of the video / audio elements with `autoplay` not load?

We don't have any video / audio elements in the app at all. Maybe, some dependency that we have (like TinyMCE) injects something at runtime but I haven't observed that. I'd have to dig more to understand.

> If you'd be willing to run some tests, here's what I'd be looking for:
> 
> * What are you using to run axe? Are you using axe-core directly or using one of our APIs (Puppeteer, Playwright, etc.)?
> * If running in a browser, I'd run axe directly using `await axe.run()` and look at the [Chrome Flame Chart](https://developer.chrome.com/docs/devtools/performance/reference/) to see if there is a particular function within axe-core that is slow. You could even save the recording and upload it for me to look at if you find it doesn't contain any private facing data. If you'd rather email me instead of post, I can give you my email in a private message.
> * I'd also run axe with our built in performance tracking `await axe.run({ performanceTimer: true })`. These just log how long parts of the code took, so you posting them here would be helpful as well.
> * If you could tell me how many audio and video elements there are on the page, as well as if there are any iframes, that'd help narrow down if it's a problem with there being lots of nodes or just a few (`document.querySelectorAll('iframe, frame').length`, `document.querySelectorAll('audio[autoplay], video[autoplay]').length`).

I'm using `axe` through [ember-a11y-testing](https://github.com/ember-a11y/ember-a11y-testing). The tests are run through QUnit. In general, it's an Ember environment. I don't have much control over it. 

When trying to run from console when the tests run I get this error:

> Axe is already running. Use `await axe.run()` to wait for the previous run to finish before starting a new run.

So I might need to setup the environment more manually. I left a note for myself to get back to it when I have a bit more time (so in the next Wiggle Week). Thanks for helping!
PS. No pressure from my side on fixing it. I think I can live with this rule disabled. Our app doesn't have video/audio assets.
